src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -1225,96 +1296,276 @@ private Type EmitFunctionCallNode(AssociativeNode node)
+                    {
+                        return false;
+                    }
+
+                    switch (args[argIndex].Kind)
+                    {
+                        case AstKind.String:
+                            var strNode = args[argIndex] as StringNode;
+                            if (p == typeof(char))
+                            {
+                                if (strNode.Value.Length != 1) return false;
+                            }
+                            break;
+                        case AstKind.Char:
+                            if (p == typeof(string)) return false;
+                            break;
+                    }
+                }
+                else if (args[argIndex] is ExprListNode exp)
                 {
-                    var argGuides = argIdent.ReplicationGuides;
-                    EmitArray(typeof(string), argGuides, (AssociativeNode gn, int gidx) =>
+                    if (!ArrayUtils.IsEnumerable(p) || p == typeof(string))
+                    {
+                        // arg is an enumerable type, param is not.
+                        return false;
+                    }
+                    var argRank = GetArgumentRank(exp);
+                    if (argRank == -1)
                     {
-                        var repGuideNode = gn as ReplicationGuideNode;
-                        EmitOpCode(OpCodes.Ldstr, repGuideNode.ToString());
-                    });
+                        // non-rectangular (jagged) array best handled by replication
+                        return false;
+                    }
+                    if (argRank != GetRank(p)) return false;
                 }
-                else
+                else if (args[argIndex] is IdentifierNode idn)
                 {
-                    // Emit an empty string array.
-                    EmitArray<string>(arrType: typeof(string), items: null, itemEmitter: null);
+                    if (idn.ReplicationGuides.Count > 0) return false;
+
+                    var t = variables[idn.Value].Item2;
+                    if (t == typeof(object)) return false;
+
+                    if (!p.Equals(t))
+                    {
+                        var argRank = GetRank(t);
+                        var paramRank = GetRank(p);
+                        if (argRank != paramRank) return false;
+
+                        // If both have rank 0, it could also be because their type info is ambiguous.
+                        if (argRank == 0 && paramRank == 0) return false;
+                    }
                 }
-            });
+                else return false;
+                argIndex++;
+            }
+            return true;
+        }
 
-            // Emit call to load the runtimeCore argument
-            EmitOpCode(OpCodes.Ldarg_3);
+        private static int GetRank(Type type)
+        {
+            return type.IsArray ? GetArrayRank(type) : GetEnumerableRank(type);
```

Comment:
only for arrays ? not enumerables ?

Reply:
For enumerables, I use `GetEnumerableRank`.